drivers/entropy/entropy_nrf5.c

```diff
@@ -63,204 +63,165 @@
+}
 
-			while (last--) {
-				*(--d) = *s++;
-			}
-		}
+#pragma GCC push_options
+#if defined(CONFIG_BT_CTLR_FAST_ENC)
+#pragma GCC optimize ("Ofast")
+#endif
+static u16_t rng_pool_get(struct rng_pool *rngp, u8_t *buf, u16_t len)
+{
+	u32_t last  = rngp->last;
+	u32_t mask  = rngp->mask;
+	u8_t *dst   = buf;
+	u32_t first, available;
+	u32_t other_read_in_progress;
+	unsigned int key;
+
+	key = irq_lock();
+	first = rngp->first_alloc;
+
+	other_read_in_progress = (rngp->first_read ^ first);
+	available = (last - first) & mask;
+	if (available < len)
+		len = available;
+
+	rngp->first_alloc = (first + len) & mask;
+	irq_unlock(key);
+
+	while (likely(len--)) {
+		*dst++ = rngp->buffer[first];
+		first = (first + 1) & mask;
+	}
 
-		rng->first = first;
+	if (likely(!other_read_in_progress)) {
+		key = irq_lock();
+		rngp->first_read = rngp->first_alloc;
+		irq_unlock(key);
 	}
 
-	if (remaining < rng->threshold) {
+	len = dst - buf;
+	available = available - len;
+	if (available <= rngp->threshold) {
 		nrf_rng_task_trigger(NRF_RNG_TASK_START);
 	}
 
-	return octets;
+	return len;
 }
 #pragma GCC pop_options
 
-static int isr(struct rand *rng, bool store)
+static int rng_pool_put(struct rng_pool *rngp, u8_t byte)
 {
-	u8_t last;
-
-	if (!rng) {
-		return -ENOBUFS;
-	}
+	u8_t first = rngp->first_read;
+	u8_t last  = rngp->last;
+	u8_t mask  = rngp->mask;
 
-	last = rng->last + 1;
-	if (last == rng->count) {
-		last = 0;
-	}
-
-	if (last == rng->first) {
-		/* this condition should not happen, but due to probable race,
-		 * new value could be generated before NRF_RNG task is stopped.
-		 */
+	/* Signal error if the ool is full. */
+	if (((last - first) & mask) == mask) {
 		return -ENOBUFS;
 	}
 
-	if (!store) {
-		return -EBUSY;
-	}
+	rngp->buffer[last] = byte;
+	rngp->last = (last + 1) & mask;
 
-	rng->rand[rng->last] = NRF_RNG->VALUE;
-	rng->last = last;
-
-	last = rng->last + 1;
-	if (last == rng->count) {
-		last = 0;
-	}
-
-	if (last == rng->first) {
-		return 0;
-	}
-
-	return -EBUSY;
+	return 0;
 }
 
-static void isr_rand(void *arg)
+static void rng_pool_init(struct rng_pool *rngp, u16_t size, u8_t threshold)
 {
-	struct device *device = arg;
-
-	if (NRF_RNG->EVENTS_VALRDY) {
-		struct entropy_nrf5_dev_data *dev_data = DEV_DATA(device);
-		int ret;
+	rngp->first_alloc = 0;
+	rngp->first_read  = 0;
+	rngp->last	  = 0;
+	rngp->mask	  = size - 1;
+	rngp->threshold	  = threshold;
+}
 
-		ret = isr((struct rand *)dev_data->isr, true);
-		if (ret != -EBUSY) {
-			ret = isr((struct rand *)dev_data->thr,
-				  (ret == -ENOBUFS));
-			k_sem_give(&dev_data->sem_sync);
-		}
+static void isr(void *arg)
+{
+	struct entropy_nrf5_dev_data *dev_data = arg;
+	int byte, ret;
 
-		NRF_RNG->EVENTS_VALRDY = 0;
+	byte = random_byte_get();
+	if (byte < 0) {
+		return;
+	}
 
-		if (ret != -EBUSY) {
+	ret = rng_pool_put((struct rng_pool *)dev_data->isr, byte);
+	if (ret < 0) {
+		ret = rng_pool_put((struct rng_pool *)dev_data->thr, byte);
+		if (ret) {
```

Comment:
These two checks of `ret` should be consistent. Now it looks like a positive value may occur and that it is acceptable in the first case and not in the second one.

Reply:
Fixed.